stan/math/prim/mat/fun/gp_exp_quad_cov.hpp

```diff
@@ -223,23 +308,315 @@ gp_exp_quad_cov(const std::vector<Eigen::Matrix<T_x1, Eigen::Dynamic, 1>> &x1,
-      x2_new = divide_columns(x2, length_scale);
+    cov = internal::gp_exp_quad_cov(x, square(sigma),
+                                    -0.5 / square(length_scale));
+    return cov;
+  }
 
-  for (size_t i = 0; i < x1_size; ++i) {
-    for (size_t j = 0; j < x2_size; ++j) {
-      cov(i, j) = sigma_sq * exp(-0.5 * squared_distance(x1_new[i], x2_new[j]));
-    }
+  if (x_size == 0)
+    return cov;
+
+  matrix_cl x_cl(x, 1, x.size());
+  check_nan(function_name, "x", x_cl);
+  matrix_cl cov_cl = gp_exp_quad_cov(x_cl, sigma, length_scale);
+  cov = from_matrix_cl(cov_cl);
+
+  return cov;
+}
+
+/**
+ * Returns a squared exponential kernel.
+ *
+ * @param x std::vector of scalars that can be used in square distance.
+ *    This function assumes each element of x is the same size.
+ * @param sigma marginal standard deviation or magnitude
+ * @param length_scale length scale
+ * @return squared distance
+ * @throw std::domain_error if sigma <= 0, l <= 0, or
+ *   x is nan or infinite
+ */
+template <>
+inline Eigen::MatrixXd gp_exp_quad_cov(const std::vector<Eigen::VectorXd> &x,
+                                       const double &sigma,
+                                       const double &length_scale) {
+  const char *function_name = "gp_exp_quad_cov";
+  check_positive(function_name, "magnitude", sigma);
+  check_positive(function_name, "length scale", length_scale);
+
+  const size_t x_size = x.size();
+  Eigen::MatrixXd cov(x_size, x_size);
+
+  if (x_size == 0)
+    return cov;
+
+  const size_t inner_x1_size = x[0].size();
+  if (static_cast<double>(x_size * x_size)
+              / opencl_context.tuning_opts().gp_exp_quad_cov_coeff1
+          + static_cast<double>((x_size + x_size + 1) * inner_x1_size)
+                / opencl_context.tuning_opts().gp_exp_quad_cov_coeff2
+      < 1) {
+    for (size_t i = 0; i < x_size; ++i)
+      for (size_t j = 0; j < x[i].size(); ++j)
+        check_not_nan("gp_exp_quad_cov", "x", x[j][i]);
+    cov = internal::gp_exp_quad_cov(x, square(sigma),
+                                    -0.5 / square(length_scale));
+    return cov;
+  }
+
+  matrix_cl x_cl(x);
+  check_nan(function_name, "x", x_cl);
+  matrix_cl cov_cl = gp_exp_quad_cov(x_cl, sigma, length_scale);
+  cov = from_matrix_cl(cov_cl);
+
+  return cov;
+}
+
+/**
+ * Returns a squared exponential kernel.
+ *
+ * @param x std::vector of Eigen vectors of scalars.
+ * @param sigma marginal standard deviation or magnitude
+ * @param length_scale std::vector length scale
+ * @return squared distance
+ * @throw std::domain_error if sigma <= 0, l <= 0, or
+ *   x is nan or infinite
+ */
+template <>
+inline Eigen::MatrixXd gp_exp_quad_cov(
+    const std::vector<Eigen::VectorXd> &x, const double &sigma,
+    const std::vector<double> &length_scale) {
+  const char *function_name = "gp_exp_quad_cov";
+  check_positive_finite(function_name, "magnitude", sigma);
+  for (size_t i = 0; i < length_scale.size(); ++i)
```

Comment:
I don't think this loop is needed. We have check functions that accept `std::vector<>`.

Reply:
I'll check, but I'm pretty sure removing this would throw for me